[CT-2411] [Bug] Ephemeral materialization not working properly with recursive CTE

### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

When performing the expansion of the referenced ephemeral models in the final query, the generated CTE are misplaced if the query starts with a `WITH RECURSIVE` statement.

### Expected Behavior

The CTE for the ephemeral models should be inserted after the `WITH RECURSIVE` statement.

### Steps To Reproduce

Let's say we have `model_1` using ephemeral materialization:
```sql
{{ config(materialized="ephemeral") }}

SELECT
  "Hello" AS greetings
```

And `model_2` using a `WITH RECURSIVE` statement and referencing `model_1`:
```sql
{{ config(materialized="table") }}

WITH RECURSIVE

  # Not actually recursive CTE.
  target AS (
    SELECT
      "world" AS target
  )

SELECT
  CONCAT(m1.greetings, ", ", t.target, " !!") AS col
FROM
  {{ ref("model_1") }} AS m1
CROSS JOIN
  target AS t

```
The generated SQL is not valid as the `model_1` CTE is inserted between the `WITH` and the `RECURSIVE` keywords:
```sql
   create or replace table `XXXX`.`XXXX`.`model_2`  as (

WITH  __dbt__cte__model_1 as (
SELECT
  "Hello" AS greetings
),RECURSIVE

  # Not actually recursive CTE.
  target AS (
    SELECT
      "world" AS target
  )

SELECT
  CONCAT(m1.greetings, ", ", t.target, " !!") AS col
FROM
  __dbt__cte__model_1 AS m1
CROSS JOIN
  target AS t
    );
```

After looking at the source code, it seems the issue comes from [this](https://github.com/dbt-labs/dbt-core/blob/70c26f5c74c340283a12f3a9270ec4c0f7135da5/core/dbt/compilation.py#L233) code that only looks for a `WITH` keyword for determining at which position to insert the CTE generated for the ephemeral models.

### Relevant log output

_No response_

### Environment

```markdown
- OS: Linux Mint 20.3
- Python: 3.11.2
- dbt: 1.4.5
```


### Which database adapter are you using with dbt?

bigquery

### Additional Context

_No response_

The following patch:

```patch
diff --git a/core/dbt/compilation.py b/core/dbt/compilation.py
index e5503af4..47c6f3fe 100644
--- a/core/dbt/compilation.py
+++ b/core/dbt/compilation.py
@@ -231,7 +231,17 @@ class Compiler:
         with_stmt = None
         for token in parsed.tokens:
             if token.is_keyword and token.normalized == "WITH":
-                with_stmt = token
+                _, next_token = parsed.token_next(
+                    parsed.token_index(token), skip_ws=True, skip_cm=True
+                )
+                if (
+                    next_token is not None
+                    and next_token.is_keyword
+                    and next_token.normalized == "RECURSIVE"
+                ):
+                    with_stmt = next_token
+                else:
+                    with_stmt = token
                 break
 
         if with_stmt is None:
```

Works perfectly, but I can't push it right now because mypy complains:

```
core/dbt/compilation.py:234: error: "Statement" has no attribute "token_next"  [attr-defined]
core/dbt/compilation.py:235: error: "Statement" has no attribute "token_index"  [attr-defined]
```

So I guess I need to figure out how to add those definitions now.

--- a/.changes/unreleased/Fixes-20230420-123221.yaml
+++ b/.changes/unreleased/Fixes-20230420-123221.yaml
@@ -1,0 +1,6 @@
+kind: Fixes
+body: Fix CTE insertion position when the model uses WITH RECURSIVE
+time: 2023-04-20T12:32:21.432848+12:00
+custom:
+  Author: willbryant
+  Issue: "7350"

--- a/core/dbt/compilation.py
+++ b/core/dbt/compilation.py
@@ -232,6 +232,10 @@ class Compiler:
         for token in parsed.tokens:
             if token.is_keyword and token.normalized == "WITH":
                 with_stmt = token
+            elif token.is_keyword and token.normalized == "RECURSIVE" and with_stmt is not None:
+                with_stmt = token
+                break
+            elif not token.is_whitespace and with_stmt is not None:
                 break
 
         if with_stmt is None:

--- a/tests/functional/compile/fixtures.py
+++ b/tests/functional/compile/fixtures.py
@@ -32,6 +32,16 @@ select {{
 }} as fun
 """
 
+with_recursive_model_sql = """
+{{ config(materialized = 'ephemeral') }}
+with recursive t(n) as (
+    select * from {{ ref('first_ephemeral_model') }}
+  union all
+    select n+1 from t where n < 100
+)
+select sum(n) from t;
+"""
+
 schema_yml = """
 version: 2
 

--- a/tests/functional/compile/test_compile.py
+++ b/tests/functional/compile/test_compile.py
@@ -9,6 +9,7 @@ from tests.functional.compile.fixtures import (
     first_ephemeral_model_sql,
     second_ephemeral_model_sql,
     third_ephemeral_model_sql,
+    with_recursive_model_sql,
     schema_yml,
     model_multiline_jinja,
 )
@@ -54,6 +55,7 @@ class TestEphemeralModels:
             "first_ephemeral_model.sql": first_ephemeral_model_sql,
             "second_ephemeral_model.sql": second_ephemeral_model_sql,
             "third_ephemeral_model.sql": third_ephemeral_model_sql,
+            "with_recursive_model.sql": with_recursive_model_sql,
         }
 
     def test_first_selector(self, project):
@@ -100,6 +102,20 @@ class TestEphemeralModels:
             ")select * from __dbt__cte__second_ephemeral_model",
             "union all",
             "select 2 as fun",
+        ]
+
+    def test_with_recursive_cte(self, project):
+        run_dbt(["compile"])
+
+        assert get_lines("with_recursive_model") == [
+            "with recursive  __dbt__cte__first_ephemeral_model as (",
+            "select 1 as fun",
+            "),t(n) as (",
+            "    select * from __dbt__cte__first_ephemeral_model",
+            "  union all",
+            "    select n+1 from t where n < 100",
+            ")",
+            "select sum(n) from t;",
         ]
 
 

--- a/third-party-stubs/sqlparse/sql.pyi
+++ b/third-party-stubs/sqlparse/sql.pyi
@@ -3,6 +3,7 @@ from typing import Tuple, Iterable
 class Token:
     def __init__(self, ttype, value): ...
     is_keyword: bool
+    is_whitespace: bool
     normalized: str
 
 class TokenList(Token):

